python/lsst/afw/table/catalog.py

```diff
@@ -88,8 +120,11 @@ def __getattribute__(self, name):
         try:
             return object.__getattribute__(self, name)
         except AttributeError:
+            # self._columns is created the when self.columns is accessed -
+            # looking for it in self.columns below would trigger infinite
+            # recursion.
```

Comment:
I think the best solution is to cache in C++. It would dramatically clean up the wrapper code with little added complexity to the C++.